planner/cascades/optimize.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cascades
+
+import (
+	plannercore "github.com/pingcap/tidb/planner/core"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pkg/errors"
+)
+
+// FindBestPlan is the optimization entrance of the cascades planner. The
+// optimization is composed of 2 phases: exploration and implementation.
+func FindBestPlan(sctx sessionctx.Context, logical plannercore.LogicalPlan) (plannercore.Plan, error) {
+	rootGroup := convert2Group(logical)
+
+	err := onPhaseExploration(sctx, rootGroup)
+	if err != nil {
+		return nil, err
+	}
+
+	best, err := onPhaseImplementation(sctx, rootGroup)
+	return best, err
+}
+
+// convert2Group converts a logical plan to expression groups.
+func convert2Group(node plannercore.LogicalPlan) *Group {
+	e := NewGroupExpr(node)
+	e.children = make([]*Group, 0, len(node.Children()))
+	for _, child := range node.Children() {
+		childGroup := convert2Group(child)
+		e.children = append(e.children, childGroup)
+	}
+	return NewGroup(e)
+}
+
+func onPhaseExploration(sctx sessionctx.Context, g *Group) error {
+	for !g.explored {
+		err := exploreGroup(g)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func exploreGroup(g *Group) error {
+	if g.explored {
+		return nil
+	}
+
+	g.explored = true
+	for elem := g.equivalents.Front(); elem != nil; elem.Next() {
+		curExpr := elem.Value.(*GroupExpr)
+		if curExpr.explored {
+			continue
+		}
+
+		// Explore child groups firstly.
+		curExpr.explored = true
+		for _, childGroup := range curExpr.children {
+			exploreGroup(childGroup)
+			curExpr.explored = curExpr.explored && childGroup.explored
+		}
+
+		eraseCur, err := findMoreEquiv(curExpr, g)
+		if err != nil {
+			return err
+		}
+		if eraseCur {
+			g.Delete(curExpr)
+		}
+
+		g.explored = g.explored && curExpr.explored
+	}
+	return nil
+}
+
+// Find and apply the matched transformation rules.
+func findMoreEquiv(expr *GroupExpr, g *Group) (eraseCur bool, err error) {
+	for _, rule := range GetTransformationRules(expr.exprNode) {
+		pattern := rule.GetPattern()
+		// Create a binding of the current group expression and the pattern of
+		// the transformation rule to enumerate all the possible expressions.
+		for iter := NewExprIter(expr, pattern); iter != nil && iter.OK(); iter.Next() {
```

Comment:
same concern as above, `ExprIter` need to handle "removal" case?

Reply:
Yes, it should.

In the context of `ExprIter`, we only concern the `operand` for each `GroupExpr` in the `Group`, I'm considering using an ordered list to store the `GroupExpr` in the `Goup`, and store the first `GroupExpr` for each `operand` in a map.
1. insert: find the first `GroupExpr` in the map, if it's not existed, use the binary search to find the insert position
2. delete: use the `exists` map to find the position of the `GroupExpr`, delete it and maintain the `operand` map.

As you can see, the `operand` map can be easily maintained, and we can quickly find the first match in the `Group`.

Reply:
Does it needs to be ordered? We just need to make sure that all the `GroupExpr` with the same `operand` is consecutive. If it's not existed, just put it in the front or end of the list.